extensions/amp-geo/0.1/amp-geo.js

```diff
@@ -0,0 +1,249 @@
+ * Example:
+ * <code>
+ * <amp-geo>
+ *  <script type="application-json">
+ *  {
+ *    ISOCountryGroups: {
+ *      "anz": [ "au", "nz" ],
+ *      "nafta": [' "ca", "mx", "us", "unknown" ],
+ *      "iceland": [ "is" ]
+ *    }
+ *  }
+ *  </scirpt>
+ * </amp-geo>
+ * </code>
+ *
+ * the amp-geo element's layout type is nodisplay.
+ */
+
+import {Layout} from '../../../src/layout';
+import {childElementsByTag, isJsonScriptTag} from '../../../src/dom';
+import {getMode} from '../../../src/mode';
+import {isArray, isObject} from '../../../src/types';
+import {isCanary} from '../../../src/experiments';
+import {isProxyOrigin} from '../../../src/url';
+import {parseJson} from '../../../src/json';
+import {registerServiceBuilder} from '../../../src/service';
+import {user} from '../../../src/log';
+import {waitForBodyPromise} from '../../../src/dom';
+
+
+/** @const */
+const TAG = 'amp-geo';
+/**
+ * COUNTRY is a special const the magic string AMP_ISO_COUNTRY_HOTPATCH
+ * is replaced at serving time with the two letter country code or
+ * error value padded to length to avoid breaking .map files.  We
+ * then trim it and store is as this.country_
+ *
+ * So don't change the magic string!
+ */
+const COUNTRY = 'AMP_ISO_COUNTRY_HOTPATCH';
+const COUNTRY_PREFIX = 'amp-iso-country-';
+const GROUP_PREFIX = 'amp-geo-group-';
+const PRE_RENDER_REGEX = new RegExp(`${COUNTRY_PREFIX}(\\w+)`);
+const GEO_ID = 'ampGeo';
+const defaultLen = COUNTRY.length;
+
+/**
+ * Operating Mode
+ * @enum {number}
+ */
+const mode = {
+  GEO_HOT_PATCH: 0, // Default mode, geo is patched by GFE when js is served
+  GEO_PRERENDER: 1, // We've been prerendered by an AMP Cache or publisher CMS
+  GEO_OVERRIDE: 2, //  We've been overriden in test by #amp-geo=xx
+};
+
+
+export class AmpGeo extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {number} */
+    this.mode_ = mode.GEO_HOT_PATCH;
+    /** @private {string} */
+    this.country_ = 'unknown';
+    /** @private {Array<string>} */
+    this.matchedGroups_ = [];
+    /** @Private {} */
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.NODISPLAY;
+  }
+
+  /** @override */
+  buildCallback() {
+    // All geo config within the amp-geo component. There will be only
+    // one single amp-geo allowed in page.
+    const scripts = childElementsByTag(this.element, 'script');
+    user().assert(scripts.length == 1,
+        `${TAG} should have exactly one <script> child`);
+    const script = scripts[0];
+    user().assert(isJsonScriptTag(script),
+        `${TAG} consent instance config should be put in a <script>` +
+        'tag with type= "application/json"');
+    const config = parseJson(script.textContent);
+
+    /** @private @const {!Promise<!Object<string, (string|boolean)>>} */
+    const geo = this.addToBody_(config);
+
+    registerServiceBuilder(this.win, 'geo', function() {
```

Comment:
If I understand this correctly: geo service is a function that returns a geo resolving promise. 
But geo service returns a data Object `{Object<string, string>}` [here]?

Can we not having a inline function for geo service. But create a class object for it. It would be easier to provide more APIs later I believe. (For example, check if a user is within the area given an area code? )

Reply:
this.addToBody_() is returning a promise stored in geo.  So the anon function is returning a promise.  Or did I misunderstand the question?

Reply:
Guess I am expecting something like
```
class Geo {
  constructor()

  getISOCountryGroups () {
    return this.addToBody_().then(geo => return geo['ISOCountryGroups'] );
  }

  isCountryIn(countryCode) {
    ...
  }
  ...
}
```

Then 
```
registerServiceBuilder(this.win, 'geo', Geo)
```

I feel it will be much easier to access the geo service in this way. As we don't need to worry about `getter` functions.

Reply:
Hmm, I see what you are saying, I built it that way to match all the existing api's in url-replacements-impl-js lets get a second opinion on this @choumx, @rsimha do you have thoughts here?